pkg/virtctl/imageupload/imageupload.go

```diff
@@ -0,0 +1,355 @@
+	file, err := os.Open(imagePath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	namespace, _, err := c.clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+
+	virtClient, err := kubecli.GetKubevirtClientFromClientConfig(c.clientConfig)
+	if err != nil {
+		return fmt.Errorf("Cannot obtain KubeVirt client: %v", err)
+	}
+
+	pvc, err := getUploadPVC(virtClient, namespace, pvcName, noCreate)
+	if err != nil {
+		return err
+	}
+
+	if pvc == nil {
+		pvc, err = createUploadPVC(virtClient, namespace, pvcName, pvcSize, storageClass, accessMode)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("PVC %s/%s created\n", namespace, pvc.Name)
+	} else {
+		fmt.Printf("Using existing PVC %s/%s\n", namespace, pvc.Name)
+	}
+
+	err = waitUploadPodRunning(virtClient, namespace, pvcName, uploadPodWaitInterval, time.Duration(uploadPodWaitSecs)*time.Second)
+	if err != nil {
+		return err
+	}
+
+	token, err := getUploadToken(virtClient.CdiClient(), namespace, pvcName)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Uploading data to %s\n", uploadProxyURL)
+
+	err = uploadData(uploadProxyURL, token, file, insecure)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Uploading %s completed successfully\n", imagePath)
+
+	return nil
+}
+
+func getHTTPClient(insecure bool) *http.Client {
+	client := &http.Client{}
+
+	if insecure {
+		client.Transport = &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+	}
+
+	return client
+}
+
+func uploadData(uploadProxyURL, token string, file *os.File, insecure bool) error {
+	url := uploadProxyURL + uploadProxyURI
+
+	fi, err := file.Stat()
+	if err != nil {
+		return err
+	}
+
+	bar := pb.New64(fi.Size()).SetUnits(pb.U_BYTES)
+	reader := bar.NewProxyReader(file)
+
+	client := httpClientCreatorFunc(insecure)
+	req, _ := http.NewRequest("POST", url, reader)
+
+	req.Header.Add("Authorization", "Bearer "+token)
+	req.Header.Add("Content-Type", "application/octet-stream")
+
+	fmt.Println()
+	bar.Start()
+
+	resp, err := client.Do(req)
+
+	bar.Finish()
+	fmt.Println()
+
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("Unexpected return value %d", resp.StatusCode)
+	}
+
+	return nil
+}
+
+func getUploadToken(client cdiClientset.Interface, namespace, name string) (string, error) {
+	request := &uploadcdiv1.UploadTokenRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "token-for-virtctl",
+		},
+		Spec: uploadcdiv1.UploadTokenRequestSpec{
+			PvcName: name,
+		},
+	}
+
+	response, err := client.UploadV1alpha1().UploadTokenRequests(namespace).Create(request)
+	if err != nil {
+		return "", err
+	}
+
+	return response.Status.Token, nil
+}
+
+func waitUploadPodRunning(client kubernetes.Interface, namespace, name string, interval, timeout time.Duration) error {
+	serviceName := "cdi-upload-" + name
+	loggedStatus := false
+
+	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
```

Comment:
Would it make sense to establish a watch on that one resource? Not sure what's cheaper though, since I never tested it regarding to resource use.

Reply:
I don't think it matters much for this use case.  But if really important, maybe do in a subsequent PR?